clippy_lints/src/dereference.rs

```diff
@@ -920,6 +931,115 @@ fn ty_contains_infer(ty: &hir::Ty<'_>) -> bool {
     v.0
 }
 
+// Checks whether:
+// * child is an expression of the form `&e` in an argument position requiring an `impl Trait`
+// * `e`'s type implements `Trait` and is copyable
+// If the conditions are met, returns `Some(Position::ImplArg(..))`; otherwise, returns `None`.
+//   The "is copyable" condition is to avoid the case where removing the `&` means `e` would have to
+// be moved, but it cannot be.
+fn needless_borrow_impl_arg_position<'tcx>(
+    cx: &LateContext<'tcx>,
+    parent: &Expr<'tcx>,
+    child_arg_index: usize,
+    child_arg_ty: Ty<'tcx>,
+    child_id: HirId,
+) -> Option<Position> {
+    let sized_trait_def_id = cx.tcx.lang_items().sized_trait();
+
+    let Some(callee_def_id) = fn_def_id(cx, parent) else { return None };
+
+    let traits = cx
+        .tcx
+        .predicates_of(callee_def_id)
+        .predicates
+        .iter()
+        .filter_map(|(predicate, _)| {
+            if_chain! {
+                if let PredicateKind::Trait(TraitPredicate {
+                    trait_ref:
+                        TraitRef {
+                            def_id: trait_def_id,
+                            substs,
+                            ..
+                        },
+                    ..
+                }) = predicate.kind().skip_binder();
+                if let Some(arg) = substs.iter().next();
+                if let GenericArgKind::Type(arg_ty) = arg.unpack();
+                if arg_ty == child_arg_ty;
+                then { Some((trait_def_id, &substs[1..])) } else { None }
+            }
+        })
+        .collect::<Vec<_>>();
+
+    // If only `Sized` was found, return.
+    if traits
+        .iter()
+        .all(|&(trait_def_id, _)| Some(trait_def_id) == sized_trait_def_id)
+    {
+        return None;
+    }
+
+    if !matches!(child_arg_ty.kind(), ty::Param(_)) {
+        return None;
+    }
+
+    // If `child_arg_ty` is a type parameter that appears in more than one place, then substituting
+    // it with `T` instead of `&T` could cause a type error.
+    if cx
+        .tcx
+        .fn_sig(callee_def_id)
+        .skip_binder()
+        .inputs_and_output
+        .iter()
+        .enumerate()
+        .any(|(i, ty)| i != child_arg_index && contains_ty(ty, child_arg_ty))
+    {
+        return None;
+    }
+
+    let check_referent = |referent| {
+        let referent_ty = cx.typeck_results().expr_ty(referent);
+
+        if !is_copy(cx, referent_ty) {
+            return false;
+        }
+
+        // * If an applicable trait is found and `referent_ty` implements it, `needless_borrow` is set to
+        //   `true`.
+        // * If an applicable trait is found that `referent_ty` does not implement, `false` is returned
+        //   immediately.
+        // * If no applicable traits are found, `needless_borrow` remains `false`.
+        let mut needless_borrow = false;
+        for &(trait_def_id, substs) in &traits {
```

Comment:
This could use `traits.iter().all(..)`. I don't see a reason to treat `Sized` specially. All the predicates must be met no matter what they are and it's already been checked that there's at least one additional predicate.